File: kotlin/src/main/kotlin/com/hedvig/hanalytics/HAnalytics.kt
package com.hedvig.hanalytics

/**
 * An app screen
 */
enum class AppScreen(val value: String) {

    OFFER("offer"),

    CLAIMS_STATUS_DETAIL("claims_status_detail"),

    CLAIM_HONOR_PLEDGE("claim_honor_pledge"),

    COMMON_CLAIM_DETAIL("common_claim_detail"),

    EMBARK("embark"),

    EMBARK_TOOLTIP("embark_tooltip"),

    MARKET_PICKER("market_picker"),

    MARKETING("marketing"),

    DATA_COLLECTION_AUTHENTICATING("data_collection_authenticating"),

    DATA_COLLECTION_CREDENTIALS("data_collection_credentials"),

    DATA_COLLECTION_FAIL("data_collection_fail"),

    DATA_COLLECTION_INTRO("data_collection_intro"),

    DATA_COLLECTION_SUCCESS("data_collection_success"),

    CONNECT_PAYMENT_ADYEN("connect_payment_adyen"),

    CONNECT_PAYMENT_TRUSTLY("connect_payment_trustly"),

    CONNECT_PAYMENT_FAILED("connect_payment_failed"),

    CONNECT_PAYMENT_SUCCESS("connect_payment_success"),

    PAYMENTS("payments"),

    CHARITY("charity"),

    CONTACT_INFO("contact_info"),

    CROSS_SELL_DETAIL("cross_sell_detail"),

    FOREVER("forever"),

    HOME("home"),

    INSURANCE_DETAIL("insurance_Detail"),

    INSURANCES("insurances"),

    MOVING_FLOW_INTRO("moving_flow_intro"),

    PROFILE("profile"),

    APP_INFORMATION("app_information"),

    APP_SETTINGS("app_settings"),

    CHAT("chat")
    ;
}

abstract class HAnalytics {
    protected abstract fun send(event: HAnalyticsEvent)
    /**
     * A screen was shown in the app
     */
    fun screenView(screen: AppScreen) {
        send(
            HAnalyticsEvent(
                name = "app_screen_view",
                properties = mapOf(
                    "screen_name" to screen.value,
                ),
            )
        )
    }
    /**
     * When a file, video, image, gif is sent in the chat
     */
    fun chatRichMessageSent() {
        send(
            HAnalyticsEvent(
                name = "chat_rich_message_sent",
                properties = mapOf(),
            )
        )
    }
    /**
     * When a text message is sent in the chat
     */
    fun chatTextMessageSent() {
        send(
            HAnalyticsEvent(
                name = "chat_text_message_sent",
                properties = mapOf(),
            )
        )
    }
    /**
     * When a claim card has been clicked on screen
     */
    fun claimCardClick(claimId: String, claimStatus: String) {
        send(
            HAnalyticsEvent(
                name = "claim_card_click",
                properties = mapOf(
                    "claim_id" to claimId,
                    "claim_status" to claimStatus,
                ),
            )
        )
    }
    /**
     * When a claim card has been shown on screen
     */
    fun claimCardVisible(claimId: String, claimStatus: String) {
        send(
            HAnalyticsEvent(
                name = "claim_card_visible",
                properties = mapOf(
                    "claim_id" to claimId,
                    "claim_status" to claimStatus,
                ),
            )
        )
    }
    /**
     * When contact chat is tapped on claim details
     */
    fun claimDetailClickOpenChat(claimId: String, claimStatus: String) {
        send(
            HAnalyticsEvent(
                name = "claim_status_detail_click_open_chat",
                properties = mapOf(
                    "claim_id" to claimId,
                    "claim_status" to claimStatus,
                ),
            )
        )
    }
    /**
     * When a claims recording has been played in the claims status screen
     */
    fun claimsDetailRecordingPlayed(claimId: String) {
        send(
            HAnalyticsEvent(
                name = "claims_detail_recording_played",
                properties = mapOf(
                    "claim_id" to claimId,
                ),
            )
        )
    }
    /**
     * When a deep link was opened
     */
    fun deepLinkOpened(type: String) {
        send(
            HAnalyticsEvent(
                name = "deep_link_opened",
                properties = mapOf(
                    "type" to type,
                ),
            )
        )
    }
    /**
     * User just logged in
     */
    fun loggedIn() {
        send(
            HAnalyticsEvent(
                name = "logged_in",
                properties = mapOf(),
            )
        )
    }
    /**
     * User just logged out
     */
    fun loggedOut() {
        send(
            HAnalyticsEvent(
                name = "logged_out",
                properties = mapOf(),
            )
        )
    }
    /**
     * A push notification was opened
     */
    fun notificationOpened(type: String) {
        send(
            HAnalyticsEvent(
                name = "notification_opened",
                properties = mapOf(
                    "type" to type,
                ),
            )
        )
    }
    /**
     * The state of notification permission:
     *   granted == true: push notifications permissions are approved
     *   granted == false: push notifications permissions are denied
     *   granted == null: push notifications permissions are not determined yet / unknown
     *
     */
    fun notificationPermission(granted: Boolean?) {
        send(
            HAnalyticsEvent(
                name = "notification_permission",
                properties = mapOf(
                    "granted" to granted,
                ),
            )
        )
    }
    /**
     * When a user clicks "Already a member? Log in" on the marketing screen
     */
    fun buttonClickMarketingLogin() {
        send(
            HAnalyticsEvent(
                name = "button_click_marketing_login",
                properties = mapOf(),
            )
        )
    }
    /**
     * When a user clicks "Get a price quote" on the marketing screen
     */
    fun buttonClickMarketingOnboard() {
        send(
            HAnalyticsEvent(
                name = "button_click_marketing_onboard",
                properties = mapOf(),
            )
        )
    }
    /**
     * When an embark flow is choosen on the choose screen
     */
    fun onboardingChooseEmbarkFlow(embarkStoryId: String) {
        send(
            HAnalyticsEvent(
                name = "onboarding_choose_embark_flow",
                properties = mapOf(
                    "embark_story_id" to embarkStoryId,
                ),
            )
        )
    }
    /**
     * When the user decided to skip data collection
     */
    fun dataCollectionSkipped(providerId: String) {
        send(
            HAnalyticsEvent(
                name = "data_collection_skipped",
                properties = mapOf(
                    "provider_id" to providerId,
                ),
            )
        )
    }
    /**
     * When a market was selected on the market picker screen
     */
    fun marketSelected(locale: String) {
        send(
            HAnalyticsEvent(
                name = "market_selected",
                properties = mapOf(
                    "locale" to locale,
                ),
            )
        )
    }
    /**
     * A payment card was shown on the home screen
     */
    fun homePaymentCardVisible() {
        send(
            HAnalyticsEvent(
                name = "home_payment_card_visible",
                properties = mapOf(),
            )
        )
    }
    /**
     * When embark does an external redirect
     */
    fun embarkExternalRedirect(location: String) {
        send(
            HAnalyticsEvent(
                name = "embark_external_redirect",
                properties = mapOf(
                    "location" to location,
                ),
            )
        )
    }
    /**
     * When embark goes back one passage
     */
    fun embarkPassageGoBack(storyName: String, passageName: String) {
        send(
            HAnalyticsEvent(
                name = "embark_passage_go_back",
                properties = mapOf(
                    "originated_from_embark_story" to storyName,
                    "passage_name" to passageName,
                ),
            )
        )
    }
    /**
     * When embark sends a tracking event
     */
    fun embarkTrack(storyName: String, eventName: String, store: Map<String, String?>) {
        send(
            HAnalyticsEvent(
                name = "embark_track",
                properties = mapOf(
                    "originated_from_embark_story" to storyName,
                    "event_name" to eventName,
                    "store" to store,
                ),
            )
        )
    }
    /**
     * When embark does a varianted offer redirect
     */
    fun embarkVariantedOfferRedirect(allIds: List<String>, selectedIds: List<String>) {
        send(
            HAnalyticsEvent(
                name = "embark_varianted_offer_redirect",
                properties = mapOf(
                    "all_ids" to allIds,
                    "selected_ids" to selectedIds,
                ),
            )
        )
    }
    /**
     * Experiment where evaluated, typically means it was shown on screen or similar
     */
    fun experimentEvaluated(name: String, variant: String) {
        send(
            HAnalyticsEvent(
                name = "experiment_evaluated",
                properties = mapOf(
                    "name" to name,
                    "variant" to variant,
                ),
            )
        )
    }
    /**
     * Experiments where loaded from server
     */
    fun experimentsLoaded(experiments: List<String>) {
        send(
            HAnalyticsEvent(
                name = "experiments_loaded",
                properties = mapOf(
                    "experiments" to experiments,
                ),
            )
        )
    }
    /**
     * When quotes are signed in the offer screen
     */
    fun quotesSigned(quoteIds: List<String>) {
        send(
            HAnalyticsEvent(
                name = "quotes_signed",
                properties = mapOf(
                    "quote_ids" to quoteIds,
                ),
                graphql = mapOf(
                    "query" to """
query QuotesSigned(${"\$"}quote_ids: [ID!]!) {
	quoteBundle(input: {
		ids: ${"\$"}quote_ids
	}) {
		quotes {
			typeOfContract
			initiatedFrom
		}
	}
}                            """.trimIndent(),
                    "selectors" to listOf(
                        mapOf(
                            "name" to "type_of_contracts",
                            "path" to "quoteBundle.quotes[*].typeOfContract | sort(@) | join(', ', @)",
                        ),
                        mapOf(
                            "name" to "initiated_from",
                            "path" to "quoteBundle.quotes[0].initiatedFrom",
                        ),
                    ),
                    "variables" to mapOf<String, Any?>(),
                ),
            )
        )
    }
    /**
     * When the user received some quotes
     */
    fun receivedQuotes(quoteIds: List<String>) {
        send(
            HAnalyticsEvent(
                name = "received_quotes",
                properties = mapOf(
                    "quote_ids" to quoteIds,
                ),
                graphql = mapOf(
                    "query" to """
query ReceivedQuotes(${"\$"}quote_ids: [ID!]!) {
	quoteBundle(input: {
		ids: ${"\$"}quote_ids
	}) {
		quotes {
			typeOfContract
			initiatedFrom
		}
	}
}                            """.trimIndent(),
                    "selectors" to listOf(
                        mapOf(
                            "name" to "type_of_contracts",
                            "path" to "quoteBundle.quotes[*].typeOfContract | sort(@)",
                        ),
                        mapOf(
                            "name" to "initiated_from",
                            "path" to "quoteBundle.quotes[0].initiatedFrom",
                        ),
                    ),
                    "variables" to mapOf<String, Any?>(),
                ),
            )
        )
    }
    /**
     * App was put into background
     */
    fun appBackground() {
        send(
            HAnalyticsEvent(
                name = "app_background",
                properties = mapOf(),
            )
        )
    }
    /**
     * App was resumed after being in background
     */
    fun appResumed() {
        send(
            HAnalyticsEvent(
                name = "app_resumed",
                properties = mapOf(),
            )
        )
    }
    /**
     * App was started
     */
    fun appStarted() {
        send(
            HAnalyticsEvent(
                name = "app_started",
                properties = mapOf(),
            )
        )
    }
}

File: swift/hAnalytics.swift
import Foundation

public struct hAnalyticsEvent {
  public let name: String
  public let properties: [String: Any?]
  public let graphql: [String: Any]?

  public init(name: String, properties: [String: Any?], graphql: [String: Any]? = nil) {
    self.name = name
    self.properties = properties
    self.graphql = graphql
  }
}

public struct hAnalyticsParcel {
  var sender: () -> Void

  init(_ sender: @escaping () -> Void) { self.sender = sender }

  /// sends the event instantly
  public func send() { sender() }
}

extension hAnalyticsEvent {
  /// identifies and registers the trackingId
  public static func identify() { hAnalyticsNetworking.identify() }

  /// When a claim card has been clicked on screen
  public static func claimCardClick(claimId: String, claimStatus: String) -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = ["claim_id": claimId, "claim_status": claimStatus]

      hAnalyticsNetworking.send(hAnalyticsEvent(name: "claim_card_click", properties: properties))
    }
  }

  /// When a claim card has been shown on screen
  public static func claimCardVisible(claimId: String, claimStatus: String) -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = ["claim_id": claimId, "claim_status": claimStatus]

      hAnalyticsNetworking.send(hAnalyticsEvent(name: "claim_card_visible", properties: properties))
    }
  }

  /// When contact chat is tapped on claim details
  public static func claimDetailClickOpenChat(claimId: String, claimStatus: String)
    -> hAnalyticsParcel
  {
    return hAnalyticsParcel {
      let properties: [String: Any?] = ["claim_id": claimId, "claim_status": claimStatus]

      hAnalyticsNetworking.send(
        hAnalyticsEvent(name: "claim_status_detail_click_open_chat", properties: properties)
      )
    }
  }

  /// When a claims recording has been played in the claims status screen
  public static func claimsDetailRecordingPlayed(claimId: String) -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = ["claim_id": claimId]

      hAnalyticsNetworking.send(
        hAnalyticsEvent(name: "claims_detail_recording_played", properties: properties)
      )
    }
  }

  /// When the claims status detail screen is shown
  public static func claimsStatusDetailScreenView(claimId: String, claimStatus: String)
    -> hAnalyticsParcel
  {
    return hAnalyticsParcel {
      let properties: [String: Any?] = ["claim_id": claimId, "claim_status": claimStatus]

      hAnalyticsNetworking.send(
        hAnalyticsEvent(name: "claims_status_detail_screen_view", properties: properties)
      )
    }
  }

  /// When a deep link was opened
  public static func deepLinkOpened(type: String) -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = ["type": type]

      hAnalyticsNetworking.send(hAnalyticsEvent(name: "deep_link_opened", properties: properties))
    }
  }

  /// When embark does an external redirect
  public static func embarkExternalRedirect(location: String) -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = ["location": location]

      hAnalyticsNetworking.send(
        hAnalyticsEvent(name: "embark_external_redirect", properties: properties)
      )
    }
  }

  /// When embark goes back one passage
  public static func embarkPassageGoBack(storyName: String, passageName: String) -> hAnalyticsParcel
  {
    return hAnalyticsParcel {
      let properties: [String: Any?] = [
        "originated_from_embark_story": storyName, "passage_name": passageName,
      ]

      hAnalyticsNetworking.send(
        hAnalyticsEvent(name: "embark_passage_go_back", properties: properties)
      )
    }
  }

  /// When embark sends a tracking event
  public static func embarkTrack(storyName: String, eventName: String, store: [String: Any])
    -> hAnalyticsParcel
  {
    return hAnalyticsParcel {
      let properties: [String: Any?] = [
        "originated_from_embark_story": storyName, "event_name": eventName, "store": store,
      ]

      hAnalyticsNetworking.send(hAnalyticsEvent(name: "embark_track", properties: properties))
    }
  }

  /// When embark does a varianted offer redirect
  public static func embarkVariantedOfferRedirect(allIds: [String], selectedIds: [String])
    -> hAnalyticsParcel
  {
    return hAnalyticsParcel {
      let properties: [String: Any?] = ["all_ids": allIds, "selected_ids": selectedIds]

      hAnalyticsNetworking.send(
        hAnalyticsEvent(name: "embark_varianted_offer_redirect", properties: properties)
      )
    }
  }

  /// When embark is shown
  public static func screenViewEmbark() -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = [:]

      hAnalyticsNetworking.send(hAnalyticsEvent(name: "screen_view_embark", properties: properties))
    }
  }

  /// When quotes are signed in the offer screen
  public static func quotesSigned(quoteIds: [String]) -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = ["quote_ids": quoteIds]

      let graphQLVariables: [String: Any?] = ["quote_ids": quoteIds]

      hAnalyticsNetworking.send(
        hAnalyticsEvent(
          name: "quotes_signed",
          properties: properties,
          graphql: [
            "query": """
              query QuotesSigned($quote_ids: [ID!]!) {
              	quoteBundle(input: {
              		ids: $quote_ids
              	}) {
              		quotes {
              			typeOfContract
              			initiatedFrom
              		}
              	}
              }
              """,
            "selectors": [
              [
                "name": "type_of_contracts",
                "path": "quoteBundle.quotes[*].typeOfContract | sort(@) | join(', ', @)",
              ], ["name": "initiated_from", "path": "quoteBundle.quotes[0].initiatedFrom"],
            ], "variables": graphQLVariables,
          ]
        )
      )
    }
  }

  /// When a user clicks &#34;Already a member? Log in&#34; on the marketing screen
  public static func buttonClickMarketingLogin() -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = [:]

      hAnalyticsNetworking.send(
        hAnalyticsEvent(name: "button_click_marketing_login", properties: properties)
      )
    }
  }

  /// When a user clicks &#34;Get a price quote&#34; on the marketing screen
  public static func buttonClickMarketingOnboard() -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = [:]

      hAnalyticsNetworking.send(
        hAnalyticsEvent(name: "button_click_marketing_onboard", properties: properties)
      )
    }
  }

  /// When an embark flow is choosen on the choose screen
  public static func onboardingChooseEmbarkFlow(embarkStoryId: String) -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = ["embark_story_id": embarkStoryId]

      hAnalyticsNetworking.send(
        hAnalyticsEvent(name: "onboarding_choose_embark_flow", properties: properties)
      )
    }
  }

  /// When the user decided to skip data collection
  public static func dataCollectionSkipped(providerId: String) -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = ["provider_id": providerId]

      hAnalyticsNetworking.send(
        hAnalyticsEvent(name: "data_collection_skipped", properties: properties)
      )
    }
  }

  /// When data collection waiting for authentication screen is shown
  public static func screenViewDataCollectionAuthenticating(providerId: String) -> hAnalyticsParcel
  {
    return hAnalyticsParcel {
      let properties: [String: Any?] = ["provider_id": providerId]

      hAnalyticsNetworking.send(
        hAnalyticsEvent(name: "screen_view_data_collection_authenticating", properties: properties)
      )
    }
  }

  /// When data collection credentials screen is shown (Insurely)
  public static func screenViewDataCollectionCredentials(providerId: String) -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = ["provider_id": providerId]

      hAnalyticsNetworking.send(
        hAnalyticsEvent(name: "screen_view_data_collection_credentials", properties: properties)
      )
    }
  }

  /// When data collection failed
  public static func screenViewDataCollectionFail(providerId: String) -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = ["provider_id": providerId]

      hAnalyticsNetworking.send(
        hAnalyticsEvent(name: "screen_view_data_collection_fail", properties: properties)
      )
    }
  }

  /// When data collection intro screen is shown (Insurely)
  public static func screenViewDataCollectionIntro(providerId: String) -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = ["provider_id": providerId]

      hAnalyticsNetworking.send(
        hAnalyticsEvent(name: "screen_view_data_collection_intro", properties: properties)
      )
    }
  }

  /// When data collection succeeded
  public static func screenViewDataCollectionSuccess(providerId: String) -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = ["provider_id": providerId]

      hAnalyticsNetworking.send(
        hAnalyticsEvent(name: "screen_view_data_collection_success", properties: properties)
      )
    }
  }

  /// When market picker is shown
  public static func screenViewMarketPicker() -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = [:]

      hAnalyticsNetworking.send(
        hAnalyticsEvent(name: "screen_view_market_picker", properties: properties)
      )
    }
  }

  /// When a market was selected on the market picker screen
  public static func marketSelected(locale: String) -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = ["locale": locale]

      hAnalyticsNetworking.send(hAnalyticsEvent(name: "market_selected", properties: properties))
    }
  }

  /// When Offer screen is shown
  public static func screenViewOffer(offerIds: [String]) -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = ["offer_ids": offerIds]

      let graphQLVariables: [String: Any?] = ["offer_ids": offerIds]

      hAnalyticsNetworking.send(
        hAnalyticsEvent(
          name: "screen_view_offer",
          properties: properties,
          graphql: [
            "query": """
              query ScreenViewOffer($offer_ids: [ID!]!) {
              	quoteBundle(input: {
              		ids: $offer_ids
              	}) {
              		quotes {
              			typeOfContract
              			initiatedFrom
              		}
              	}
              }
              """,
            "selectors": [
              [
                "name": "type_of_contracts",
                "path": "quoteBundle.quotes[*].typeOfContract | sort(@) | join(', ', @)",
              ], ["name": "initiated_from", "path": "quoteBundle.quotes[0].initiatedFrom"],
            ], "variables": graphQLVariables,
          ]
        )
      )
    }
  }

  /// When marketing screen is shown
  public static func screenViewMarketing() -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = [:]

      hAnalyticsNetworking.send(
        hAnalyticsEvent(name: "screen_view_marketing", properties: properties)
      )
    }
  }

  /// A payment card was shown on the home screen
  public static func homePaymentCardVisible() -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = [:]

      hAnalyticsNetworking.send(
        hAnalyticsEvent(name: "home_payment_card_visible", properties: properties)
      )
    }
  }

  /// Connecting payment with Adyen screen was shown
  public static func screenViewConnectPaymentAdyen() -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = [:]

      hAnalyticsNetworking.send(
        hAnalyticsEvent(name: "screen_view_connect_payment_adyen", properties: properties)
      )
    }
  }

  /// When payment connection did fail
  public static func screenViewConnectPaymentFailed() -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = [:]

      hAnalyticsNetworking.send(
        hAnalyticsEvent(name: "screen_view_connect_payment_failed", properties: properties)
      )
    }
  }

  /// When payment was connected successfully
  public static func screenViewConnectPaymentSuccess() -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = [:]

      hAnalyticsNetworking.send(
        hAnalyticsEvent(name: "screen_view_connect_payment_success", properties: properties)
      )
    }
  }

  /// Connecting payment with Trustly screen was shown
  public static func screenViewConnectPaymentTrustly() -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = [:]

      hAnalyticsNetworking.send(
        hAnalyticsEvent(name: "screen_view_connect_payment_trustly", properties: properties)
      )
    }
  }

  /// When the detail screen of a cross sell is shown
  public static func screenViewCrossSellDetail(typeOfContract: String) -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = ["type_of_contract": typeOfContract]

      hAnalyticsNetworking.send(
        hAnalyticsEvent(name: "screen_view_cross_sell_detail", properties: properties)
      )
    }
  }

  /// When Hedvig Forever is shown
  public static func screenViewForever() -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = [:]

      hAnalyticsNetworking.send(
        hAnalyticsEvent(name: "screen_view_forever", properties: properties)
      )
    }
  }

  /// When Home tab is shown
  public static func screenViewHome() -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = [:]

      hAnalyticsNetworking.send(hAnalyticsEvent(name: "screen_view_home", properties: properties))
    }
  }

  /// When detail screen of an insurance is shown
  public static func screenViewInsuranceDetail(contractId: String) -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = ["contract_id": contractId]

      hAnalyticsNetworking.send(
        hAnalyticsEvent(name: "screen_view_insurance_detail", properties: properties)
      )
    }
  }

  /// When Insurances tab is shown
  public static func screenViewInsurances() -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = [:]

      let graphQLVariables: [String: Any?] = [:]

      hAnalyticsNetworking.send(
        hAnalyticsEvent(
          name: "screen_view_insurances",
          properties: properties,
          graphql: [
            "query": """
              query ScreenViewInsurances {
              	contracts {
              		typeOfContract
              	}
              }
              """,
            "selectors": [
              [
                "name": "has_accident_insurance",
                "path":
                  "(contracts[?contains(typeOfContract, 'ACCIDENT') == `true`] && true) == true",
              ],
              [
                "name": "has_home_insurance",
                "path":
                  "((contracts[?contains(typeOfContract, 'HOME') == `true`] || contracts[?contains(typeOfContract, 'APARTMENT') == `true`] || contracts[?contains(typeOfContract, 'HOUSE') == `true`]) && true) == true",
              ],
            ], "variables": graphQLVariables,
          ]
        )
      )
    }
  }

  /// When moving flow intro screen is shown
  public static func screenViewMovingFlowIntro() -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = [:]

      hAnalyticsNetworking.send(
        hAnalyticsEvent(name: "screen_view_moving_flow_intro", properties: properties)
      )
    }
  }

  /// When Profile tab is shown
  public static func screenViewProfile() -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = [:]

      hAnalyticsNetworking.send(
        hAnalyticsEvent(name: "screen_view_profile", properties: properties)
      )
    }
  }

  /// App was put into background
  public static func appBackground() -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = [:]

      hAnalyticsNetworking.send(hAnalyticsEvent(name: "app_background", properties: properties))
    }
  }

  /// App was resumed after being in background
  public static func appResumed() -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = [:]

      hAnalyticsNetworking.send(hAnalyticsEvent(name: "app_resumed", properties: properties))
    }
  }

  /// App was shutdown
  public static func appShutdown() -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = [:]

      hAnalyticsNetworking.send(hAnalyticsEvent(name: "app_shutdown", properties: properties))
    }
  }

  /// App was started
  public static func appStarted() -> hAnalyticsParcel {
    return hAnalyticsParcel {
      let properties: [String: Any?] = [:]

      hAnalyticsNetworking.send(hAnalyticsEvent(name: "app_started", properties: properties))
    }
  }

}
